### ISSUE TITLE
Integrating implementations of SRFIs

### ISSUE THREAD

**Issue by purefunctor** [purescm/purescm#131]
SRFIs (https://srfi.schemers.org/) or Scheme Requests for Implementation extend the base language, sort of like a minimalist standard library.

SRFIs like 125 (https://srfi.schemers.org/srfi-125/srfi-125.html) allow for a richer hash table API than that offered by R6RS. Likewise, SRFI 214 (https://srfi.schemers.org/srfi-214/srfi-214.html) is useful as it implements an `ArrayList`.

[chez-srfi](https://github.com/arcfide/chez-srfi) provides implementations of SRFIs in Chez Scheme.

How should we integrate `chez-srfi` into the project?
1. We could vendor them by storing them as strings in purs files
2. We could download them during build time and copy them to the output folder
3. We could have a soft-dependency to [akku](https://akkuscm.org/)

**Comment by JordanMartinez** [comment]
I'm for vendoring those in.

**Comment by purefunctor** [comment]
The other option for vendoring is by having a `vendor` folder in the project that gets:
1. Bundled together with the `npm` package and then copied to the `output` folder
2. Pointed to with the `--libdirs` flag when running `purescm`'s test suite

`chez-srfi` has a specific file structure and laying it out from PS strings is a bit more involved than just having a local copy of it. That being said, we'd have to solve this again when we compile `purescm` to Scheme

**Comment by purefunctor** [comment]
Circling back to this, I think the problem we're ultimately trying to solve is figuring out how to represent records and arrays. It just so happens that SRFI 125 / SRFI 214 describes what we need that isn't inherently provided by the language. My main argument with vendoring chez-srfi as a whole pivots on the hypothetical that is "what if foreign libraries we bind to have to use chez-srfi? then there would be breakage"; though, the more I think about it, I don't think there's much of "foreign" libraries to benefit from in the first place, at least when you compare it to other languages.

Scheme is already sparse as-is, and in our case, it doesn't really help that Chez has only been recently open-sourced and most of the Schemes with richer ecosystems (like Racket) tend to be on the heavier side.

Taking a page from how Idris2 handles [FFI](https://idris2.readthedocs.io/en/latest/ffi/ffi.html), they just go straight to native, where the compiler generates the necessary code for:
1. loading the shared library
2. writing bindings for foreign functions

With this in mind, the relationship that Idris2 has with Scheme is more of the latter being used as a runtime for gluing stuff together rather than a transpilation target; PureScript does the latter where FFI comes in pairs of `.purs` and `.scm` files.

To summarize my thoughts on the matter:
1. Scheme's library situation is not as rich as with Node/Web+NPM, so vendoring `chez-srfi` as a whole to avoid breakage doesn't hold much ground if there are not many libraries to benefit from.
2. PureScript FFI works best when you have a rich ecosystem of packages available for the backend language like with JS. With Scheme, foreign files would likely be used exclusively for primitive operations rather than binding against existing libraries. It also subverts the process of writing FFI into native libraries as it's a bit more involved.

**Comment by JordanMartinez** [comment]
Good points raised above.

My main argument for only vendoring in what we need (as that was what I originally understood this issue being about, not all SRFIs) is that we need it now to make progress. Vendoring in the remaining SRFIs seems a bit heavy-handed when we're not sure that's even needed yet. Once we've implemented FFI for the remaining `core` libraries and probably some `contrib` libraries, I imagine we'd have a better idea of the following:
- whether or not vendoring in more SRFIs (or all of it) is even worth doing. 
- the kinds of problems that arise when bundling the `purescm`-outputted Chez Scheme code into a single executables that work on all major OSes
- how `akku` integrates into things

We can follow Idris in defining some other spec that enables one to define FFI similar to Idris, which exists outside of the PureScript compiler. For example, the `.purs` file will still using `foreign import`, but `purescm` may include `.chez-scheme-ffi` files that look like this:
```
depends-on: some-ffi-library
imports: something
exports: something
```

And `purescm` could generate a scheme file that is imported by the file corresponding to the `.purs` file.

But all of the above can be figured out once we need to figure that out.

**Comment by JordanMartinez** [comment]
> I'll create a new issue to track FFI semantics, but an idea that came to mind is that we can make use of directives to define FFI, something like:

Sounds like another case for annotation syntax in PureScript.
